src/rulesets/Base/array.jl

```diff
@@ -278,3 +278,128 @@ function rrule(::typeof(fill), value::Any, dims::Int...)
     end
     return fill(value, dims), fill_pullback
 end
+
+#####
+##### `findmax`, `maximum`, etc.
+#####
+
+for findm in (:findmin, :findmax)
+    findm_pullback = Symbol(findm, :_pullback)
+
+    @eval function frule((_, xdot), ::typeof($findm), x; dims=:)
+        y, ind = $findm(x; dims=dims)
+        return (y, ind), Tangent{typeof((y, ind))}(xdot[ind], NoTangent())
+    end
+
+    @eval function rrule(::typeof($findm), x::AbstractArray{<:Number}; dims=:)
+        y, ind = $findm(x; dims=dims)
+        project = ProjectTo(x)
+        function $findm_pullback((dy, _))
+            x_thunk = @thunk begin
+                dx = fill!(similar(x, eltype(dy)), false)
+                view(dx, ind) .= dy  # possibly 0-dim view, allows dy::Number and dy::Array, and dx::CuArray
+                project(dx)
+            end
+            x_ithunk = InplaceableThunk(x_thunk) do dx
+                view(dx, ind) .= view(dx, ind) .+ dy  # this could be .+=, but not on Julia 1.0
+                dx
+            end
+            return (NoTangent(), x_ithunk)
+        end
+        function $findm_pullback(::Tuple{AbstractZero, Any})
+            return (NoTangent(), NoTangent())
+        end
+        return (y, ind), $findm_pullback
+    end
+
+end
+
+# These rules for maximum pick the same subgradient as findmax:
+
+function frule((_, xdot), ::typeof(maximum), x; dims=:)
+    y, ind = findmax(x; dims=dims)
+    return y, xdot[ind]
+end
+
+function rrule(::typeof(maximum), x::AbstractArray{<:Number}; dims=:)
+    (y, _), back = rrule(findmax, x; dims=dims)
+    maximum_pullback(dy) = back((dy, nothing))
```

Comment:
This is my confusion about Tangent. Why shouldn't this be `Tangent{typeof((y, 1))}(dy, NoTangent())`? Or if it's not important to keep the forward type, then why not just `Tangent(dy, NoTangent())`, or `Tangent{Tuple}(dy, NoTangent())`?

Reply:
Yeah, actually both `Tangent{Tuple}(dy, NoTangent())` and `Tangent{typeof((y, 1))}(dy, NoTangent())` are correct (and my suggestion is not) because `Tangent` is parameterised by the primal type it represents the tangent for. And the tangent type is (usually) inferred from the constructor.

`Tangent(dy, NoTangent())` does not work at present though. I think I wanted to have this constructor at some point (but the motivation was mostly for easier testing)

Reply:
Ok. Since it's calling a function defined a few lines up in the same file, my vote is not to construct any fancy types -- they are friends so needn't go through the front door.